### ISSUE TITLE
Add semantics to logs so build step UI can improve log usability and visibility

### ISSUE THREAD

**Issue by sphet** [buildbot/buildbot#3538]
We have artists, designers and engineers looking at the build page. In BuildBot 0.8 we did a lot of work to provide additional insight for the results of a builder, but now that we have moved to 0.9 we want to do something as part of the head repository. Our hope is to add some semantics to the logs and a heuristic to the 'global' build step UI to improve it. However, we don't want to go down a route that won't be pulled into the head since our goal is to avoid maintaining our own branch.

We would like to have the UI, regardless of step result, always open a log that is marked as 'important' or 'summary'. For a warning or failing build, a log marked as 'errors' would always be open, regardless of how many logs are attached to the step. Should an error log not exist, stdio would always be opened in the event of a failure/warning. In this way we can provide summaries to buildsteps that are always visible and do not require the user to open them. We can also include additional logs for build engineers but they won't clutter up the user interface by being opened unless needed.

We can implement this in one of two ways:

1. Simply adopt a naming convention for the logs: errors and summary
2. Support semantics inside the data API and expose those semantics through json and data queries, much like the text or html encoding is done.

I'm in favour of either approach, but would want to implement this feature in a way that people find useful and that has a higher chance of being pulled into the head repository.

Can I get some feedback on this?

**Comment by tardyp** [comment]
Obviously, option 1 is much simpler to implement and would allow you to experiment on it, leaving the space to implement option 2 in the long term.

I wonder if log is the right place to implement that summary.
- Logs are still in the middle of the build page, which can be annoying to scroll to. 
- Logs are far away in the data model tree, and thus requires quite a lot of API calls in order to access them.

There is already the build summary mechanism that we have, which concatenates important summary to the steps, and forward that at the build resource level (state_string).
This means that a dashboard only has to download the build ressource to get access to the summary.
The state_string is defined as a Text in SQL, so that it can be arbitrarily long. Nothing would prevent us to make it markdown formatted in order to add a bit more formatting capability to that data.

Another option is to use the same idea, but add another 'long_summary' item in the build ressource, which could contain the build errors.

http://docs.buildbot.net/latest/developer/cls-buildsteps.html#buildbot.process.buildstep.BuildStep.getCurrentSummary

**Comment by sphet** [comment]
@tardyp - First thank you for the reply. I agree that option 1 would be the easiest. One of the other engineers also discussed whether logs are the right place as well. He also wondered if a summary mechanism at the top of the build steps would be better. However, we also both acknowledged that the amount of information we would want to present would be too great -- 20 lines of text say -- and at times might be coming directly from a log file. Furthermore, I feel losing the context of which step generated the data would be confusing when multiple steps might want to report summary information which includes mark down.  I also appreciate that logs are shown for each step even if the build is still not yet complete.

I am open to either as all I really want is a consistent conduit to get data in front of our artists in the easiest way. Right now our summary data is a mix of html and plaintext but we would like to use markdown so that summary can be tinted red/green for fail/success, for example, and to include links and general formatting.

Perhaps a long_summary option might be fair - it could simply be added to the dictionaries returned by the steps right now and formatted appropriately. Keeping the existing summary concise has value. Does the summary mechanism work while the build is in flight? I suppose that's why there are getCurrentSummary and getResultSummary. 

I suspect we will try the log-file naming convention first since it seems the easiest to do however any other input would be great.
S
